### DISCUSSION TITLE
(Legacy) Diblis MOS Character Shader 1.2

### DISCUSSION THREAD

**Discussion by dibli-goost** [bnpr/Malt#483] [Showcase]
**Latest Update:** https://github.com/bnpr/Malt/discussions/483#discussioncomment-6061967

![Monke](https://user-images.githubusercontent.com/31767887/220259240-0c86f03f-68c2-4465-b293-59b424d3aeed.png)

This is the Malt half of my current rendering setup. I'm posting it here, feel free to do what you want with it. In order for it to work, you need to be on the latest version of the Node Groups branch found [here](https://github.com/bnpr/Malt/releases/tag/node-groups-latest).

=== Included Stuff ===

          + A really Big Shader

                   - Shadow Texture
                   - Matcap
                   - Soft Diffuse
                   - Toon Diffuse
                   - Ambient Occlusion
                   - Soft Specular
                   - Toon Specular
                   - Rimlight
                   - Vertex Color Normals Setup

          + A rendering tree for.. rendering.

                    - A very rudimentary bloom thingy that I need to get around to replacing at some point.
                    - A full Line Art set up for rendering Character and Background Lines.

[Diblis MOS Malt Shader.zip](https://github.com/bnpr/Malt/files/10790030/Diblis.MOS.Malt.Shader.zip)

I'll update this post with the other stuff I use for Character rendering in Malt, after I get finished remaking it all for efficiencies sake.

Let me know if the blend works, and if you like it... or if you hate it, I don't really mind :smiley:. If you have any questions, ask me! I probably don't know the answer, but maybe I do!

**Comment by dibli-goost** [comment]
Little update

![image](https://user-images.githubusercontent.com/31767887/220747221-20e8beb4-dc81-4ab4-851f-81a49112b406.png)


[Diblis MOS Malt Shader 1.1.zip](https://github.com/bnpr/Malt/files/10807635/Diblis.MOS.Malt.Shader.1.1.zip)

          * Texturing Node Group
                   - Image Texture
                   - Overlay
                   - Alpha Stuff

          * Texture Overlay Node Group
                   - This is the same overlay functionality that is found on the Texturing node, I just put it into it's own little node group because sometimes I want to overlay something on top of a non-texture.
                   - Alpha Fix simply adds the alpha channel to everything else, if the overlay image has transparency it can be weird. So that's why it's there.

          * Hair Shader Node Group
                   - Really this is just a dedicated Anisotropy and Rimlight thing, but I use this for Hair. My characters have these colored bands in their hair so yeah. I could have put this on the big shader instead, but I had it separated before for some reason, so I'm just gonna keep it separated just in case.

          * Color Mask Node Group
                   - Simply splits an input into it's respective RGBA channels, for use with masking. Has mix factors on it so that you can mess with the mix of each channel if you need to (I will). There's also an invert Alpha button, since the default Alpha value is 1 rather than 0 for most things.

**NOTE:** I should have mentioned this before, but be wary of using specular with custom normals. As of right now there's something up with the NPR Specular node and it doesn't quite interpret them correctly. The old Specular model which I've included in the Character Shader Node Group doesn't have this problem, it just doesn't have some of the parameters the new one has. You can see more about this in this discussion thread. #400

There is also an included function there which you can drop into a plugin and to get the old model with the new parameters, I completely forgot about this, and will personally replace it whenever I can get the plugin working on my end :smile:

**Comment by pragma37** [comment]
Whoah! That's some heavy use of nodes there!

This is what I get by default, with doesn't look like the images you posted:
![imagen](https://user-images.githubusercontent.com/36610999/221370497-a8692b86-dba8-4761-a8e5-916aee92a191.png)

It looks like the lack of lines was caused by the `Line Width Mask` texture.
In my PC (Nvidia) the alpha is always 0, so the line width is always multiplied by 0.
Since the output from the material is a single channel, I guess the conversion from float to vec4 could depend on the driver implementation?
Multiplying only by the first channel fixed it:
![imagen](https://user-images.githubusercontent.com/36610999/221370699-a8584856-35ac-4abd-b014-847fd6bb7301.png)
So, I guess the lesson here is to only use the channels you actually have created, just in case.

The skin shading looks broken as well, but I have not figured out the reason yet.

**Reply by pragma37** [reply]
It looks like an issue with the custom normals.
![imagen](https://user-images.githubusercontent.com/36610999/221371938-22e92829-6597-4caa-afb9-186547de7e3a.png)

**Reply by pragma37** [reply]
Yep, they're missing:
![imagen](https://user-images.githubusercontent.com/36610999/221372203-b048478e-0c07-445e-8646-3505a142bcdb.png)

**Reply by dibli-goost** [reply]
Yeah I didn't actually think to include any Custom Normals on the head for whatever reason... sorry 🥲. It should work as long as you pass some into that attribute though. In the future, I might add in my Custom Normals geonodes setup, but I need to clean it up first.

Regarding the Line Width Mask Texture, that does seem odd. I was personally under the assumption that it would work by making all vec4 channels equal the float's value. I went ahead and added in a seperate node to only grab the first channel like you said, shouldn't be anymore problems there.

----------------------

Actually, speaking of line art, currently I have a blank alpha texture on the two Line Render nodes in the Character Render Layer node tree which output to the Character Lines and Background Lines Textures respectively. I have it there so that the lines can be rendered over a transparent background. But tbh, it seems a little hacky to have to create a blank alpha texture to do this (unless if I'm missing something somewhere). Wouldn't it maybe make more sense for the default input for the color be one that's transparent? Is the driver determining the default here, so that on Nvida cards it says alpha = 0, while on AMD it says alpha = 1?

I guess I could also just make a screen pass which outputs only transparency, and then just plug that in too.

**Reply by pragma37** [reply]
It turns out that the issue with the normals was likely another driver implementation difference.
You were normalizing the custom normals, but normalizing a vector of length 0 will result in a division by 0, which results in an infinite vector, and the mix function was not working correctly in that case.

The good news is that I've made some updates in Malt to avoid these kinds of implementation detail differences.
This is how your unmodified scene looks like now on my computer:
![imagen](https://user-images.githubusercontent.com/36610999/221420421-ccf950e8-652d-4426-bbdb-a1fcf23b267a.png)

I'm sure there are other issues like this one waiting to be found, but at least these ones are already fixed. 😁

>  I have it there so that the lines can be rendered over a transparent background. But tbh, it seems a little hacky to have to create a blank alpha texture to do this (unless if I'm missing something somewhere). Wouldn't it maybe make more sense for the default input for the color be one that's transparent? Is the driver determining the default here, so that on Nvida cards it says alpha = 0, while on AMD it says alpha = 1?

Yeah, I've seen myself using that workaround too. AFAIK, when reading an empty texture, the value should always be black with alpha 1.
The only alternative would be for Malt to create the dummy transparent texture for you, which may be a good idea, actually.

> I guess I could also just make a screen pass which outputs only transparency, and then just plug that in too.

Actually, that would be slower.

**Comment by dibli-goost** [comment]
Update 2: 

```
* Diblis MoS VFX Time Movement
    - Used for distorting the location of vectors based on time. Has some settings for enabling X, Y, and Z individually.

* Noise and Voronoi Distortion Node Trees
    - Used for distorting location vectors based on procedural textures.

* DA Fire VFX Node Group
    - Outputs a distorted vector to make fire-esque effects.
    - Comes with VFX Time Movement and Noise Distortion bundled in.

* Screen March UV VFX Node Group
    - This is for raymarching weirdness. Specifically for distorting Image Textures. 
    - Malt actually already has a Raymarching implementation, but it's only accessible via code, and I'm not entirely sure how it works as I don't really have a lot of experience with GLSL. At some point I may try and figure it out.

* DA Glitch Texture VFX Node Group
    - Another Image Texture group, with the ramps 'n stuff included. 
    - DA stands for Demonicaic Ability.
    - Comes with Screen March bundled in.

```

In the zip file there is now a Malt plugin you'll want to add to your plugins directory. Within is a new Specular Type added to the NPR Specular node to circumvent the issue with custom normals. I've also tried my hand at creating a Kawase Blur node... really I just converted the code found in this PDF. (https://community.arm.com/cfs-file/__key/communityserver-blogs-components-weblogfiles/00-00-00-20-66/siggraph2015_2D00_mmg_2D00_marius_2D00_notes.pdf)

It seems to work fine for me. But I don't doubt there's a more efficient way of doing this in Malt. You can find the Kawase Node in the filter category. It needs multiple texture passes to work, so Screen Pass only!

Having added the Kawase Blur to create some more performant Bloom effects, I've removed some of the old Bloom outputs on the Main Pass as they aren't needed anymore. I've kept the Bloom Fac, but it's not currently doing anything. Not sure how I want to go about using it in the set up just yet.

If you have any issues. Let me know!

[Diblis MOS Malt Shader - 1.2.zip](https://github.com/bnpr/Malt/files/11626386/Diblis.MOS.Malt.Shader.-.1.2.zip)

**Reply by pragma37** [reply]
Always happy to see artist give a try at GLSL! 🙂

There's already a `RESOLUTION` global variable, so there's no need to put it in the function inputs.

Also, a half pixel offset would be `0.5 / RESOLUTION`. 
`0.5 / (RESOLUTION / 2.0)` is actually the same as `1.0 / RESOLUTION`, which would be a full pixel offset.

**Reply by dibli-goost** [reply]
Ah I see, I missed those. I'll fix them right away.

I've been mucking around with GLSL for awhile, but still don't quite understand all of it. I was recently taking a look at the [SDF shaders](https://github.com/bnpr/Malt/blob/dfdb64d371fadb2170f7ffc4bcc58bf301736bc9/Malt/Shaders/SDF/SDF.glsl), since I was doing the Raymarching thing with the VFX.

I don't really understand the code, but I was able to get the nodes working in Malt and tried messing around with them. I want to see if there's some efficient way to define SceneSDF in the node tree itself. Some way to arbitrarily input an SDF into the node, and it do it's thing. Just because, it seems to me SceneSDF is rather contextual to whatever your "scene" is. You need to be able to combine a bunch of primitives and noise, and whatever really. Feels more suited to a node interface, so that you can just drag and drop what you need in the node tree..

After that point, some way of actually grabbing the viewport itself as a "scene" would be interesting. Whenever I feel like going back to messing around with my shader setup, I'll try and explore the Raymarching stuff a little further.
